indico/web/client/js/react/components/WTFRemoteSearchDropdown.jsx

```diff
@@ -0,0 +1,280 @@
+    searchQuery.length < minTriggerLength && (!allowById || getIdFromQuery(searchQuery) === null);
+
+  const filterOptions = (opts, query) => {
+    if (searchDisabled) {
+      return [];
+    }
+
+    query = query.toLowerCase();
+    const id = getIdFromQuery(query);
+    return opts
+      .filter(opt => (id === null ? opt.search.includes(query) : +opt['friendly-id'] === +id))
+      .map(opt => ({
+        ...opt,
+        // this might lead to some unexpected results when
+        // searchField !== labelField, because the filtering is done
+        // on searchField but highlighting happens on labelField
+        text: highlightSearch(opt.text, query),
+      }));
+  };
+
+  const transformOptions = useCallback(
+    data => {
+      if (!Array.isArray(data)) {
+        return [];
+      }
+
+      const opts = data.map(item => ({
+        'key': item[valueField],
+        'value': item[valueField],
+        'text': item[labelField],
+        // extra values passed to <Dropdown.Item>
+        'search': item[searchField].toLowerCase(),
+        'friendly-id': item.friendly_id,
+      }));
+      return _.sortBy(opts, 'value');
+    },
+    [valueField, labelField, searchField]
+  );
+
+  const fetchData = useCallback(
+    async (params = {}) => {
+      if (!searchUrl) {
+        return;
+      }
+
+      if (searchPayload) {
+        params = {...params, ...searchPayload};
+      }
+
+      let response;
+      try {
+        setLoading(true);
+        response = await debounce(() =>
+          indicoAxios({
+            url: searchUrl,
+            method: searchMethod,
+            params,
+          })
+        );
+      } catch (error) {
+        handleAxiosError(error);
+        return;
+      } finally {
+        setLoading(false);
+      }
+
+      setOptions(transformOptions(response.data));
+    },
+    [searchUrl, searchMethod, searchPayload, transformOptions, debounce]
+  );
+
+  useEffect(() => {
+    if (preload) {
+      fetchData();
+    }
+  }, [preload, fetchData]);
+
+  useEffect(() => {
+    if (preload) {
+      return;
+    }
+
+    let params;
+    const id = getIdFromQuery(searchQuery);
+    if (allowById && id !== null) {
+      params = {id};
+    } else if (searchQuery.length >= minTriggerLength) {
+      params = {q: searchQuery};
+    }
+
+    if (params) {
+      fetchData(params);
+    }
+  }, [preload, allowById, minTriggerLength, searchQuery, fetchData]);
+
+  const onChange = (evt, {value: newValue}) => {
+    setSearchQuery('');
+    onChangeFromProps(newValue);
+    if (!isControlled) {
+      setInternalValue(newValue);
+    }
+  };
+
+  const onSearch = (evt, {searchQuery: newSearchQuery}) => {
+    setSearchQuery(newSearchQuery);
+  };
+
+  const searchDisabledMessage =
+    minTriggerLength === 0 ? (
+      undefined
+    ) : allowById ? (
+      <PluralTranslate count={minTriggerLength}>
+        <Singular>Search requires at least one character or an id prefixed with #</Singular>
```

Comment:
It's a good idea to use the count param in singular as well instead of writing "one" - languages like chinese have only one form (ie no separate plural form) and IIRC it causes some issues if the singular form doesn't contain the placeholder